walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
+import java.util.Collections;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
+  private final static int totalMinutes = 1440;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    //remove events with no requested attendees
+    HashSet<String> requested = new HashSet<>(request.getAttendees());
+    HashSet<Event> eventsSet = new HashSet<>(events);
+
+    ArrayList<TimeRange> mandatoryTimes = eventsSet
+      .stream()
+      .filter(event -> this.checkAttendees(event, request, false) == true)
+      .map(event -> event.getWhen())
+      .collect(Collectors.toCollection(ArrayList::new));
+
+    ArrayList<TimeRange> optionalTimes = eventsSet
+      .stream()
+      .filter(event -> this.checkAttendees(event, request, true) == true)
+      .map(event -> event.getWhen())
+      .collect(Collectors.toCollection(ArrayList::new));
+
+    System.out.println(mandatoryTimes);
+    System.out.println(optionalTimes);
+
+    //sort time ranges by start
+    Collections.sort(mandatoryTimes, TimeRange.ORDER_BY_START);
+    Collections.sort(optionalTimes, TimeRange.ORDER_BY_START);
+
+    ArrayList<TimeRange> mandatoryFreeTimes = findFreeTimes(mandatoryTimes, request);
+    ArrayList<TimeRange> optionalFreeTimes = findFreeTimes(optionalTimes, request);
+
+    if (optionalFreeTimes.isEmpty()) {
+      return mandatoryFreeTimes;
+    }
+    
+    return optionalFreeTimes;
+  }
+
+  private static boolean checkAttendees(Event event, MeetingRequest request, boolean checkOptional) {
+    HashSet<String> attendees = new HashSet<>(event.getAttendees());
+    HashSet<String> requested = new HashSet<>(request.getAttendees());
+    HashSet<String> optional = new HashSet<>(request.getOptionalAttendees());
+
+    for (String person: requested) {
+      if (attendees.contains(person)) {
+        return true;
+      }
+    }
+    
+    if (checkOptional) {
+      for (String person: optional) {
+        if (attendees.contains(person))
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static ArrayList<TimeRange> findFreeTimes(ArrayList<TimeRange> times, MeetingRequest request) {
+    //get free times
+    ArrayList<TimeRange> freeTimes = new ArrayList();
+    
+    if (times.isEmpty() && (request.getDuration() <= totalMinutes)) {
+      freeTimes.add(TimeRange.WHOLE_DAY);
+      return freeTimes;
+    }
+
+    if (times.isEmpty()) {
+      return freeTimes;
+    }
+
+    TimeRange intersection = times.get(0);
+
+    for (TimeRange range: times) {
+      if (intersection.contains(range)) {
+        continue;
+      } else {
+        if (intersection.overlaps(range)) {
+          intersection = new TimeRange(intersection.start(), range.end() - intersection.start());
```

Comment:
you could use [`fromStartEnd`](https://github.com/semiluna/step2020/blob/49d1ad347eada7a717aff991539595f0af93a032/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java#L176) instead of making the constructor public.